ballerina/main.bal

```diff
@@ -124,25 +139,79 @@ isolated function parseResponseAsJson(string resp) returns json|error {
+    if !isOriginallyJsonObject {
+        map<json> respContent = check resp.fromJsonStringWithType();
+        anydata|error result = trap respContent[RESULT].fromJsonWithType(expectedResponseTypedesc);
+        if result is error {
+            return handleParseResponseError(result);
+        }
+        return result;
+    }
+
+    anydata|error result = resp.fromJsonStringWithType(expectedResponseTypedesc);
     if result is error {
         return handleParseResponseError(result);
     }
     return result;
 }
 
 isolated function handleParseResponseError(error chatResponseError) returns error {
-    if chatResponseError.message().includes(JSON_CONVERSION_ERROR) 
+    if chatResponseError.message().includes(JSON_CONVERSION_ERROR)
             || chatResponseError.message().includes(CONVERSION_ERROR) {
         return error(string `${ERROR_MESSAGE}`, detail = chatResponseError);
     }
     return chatResponseError;
 }
 
-isolated function getExpectedResponseSchema(typedesc<anydata> expectedResponseTypedesc) returns map<json> {
+isolated function getExpectedResponseSchema(typedesc<anydata> expectedResponseTypedesc) returns SchemaResponse {
     // Restricted at compile-time for now.
     typedesc<json> td = checkpanic expectedResponseTypedesc.ensureType();
-    return generateJsonSchemaForTypedescAsJson(td);
+    return generateJsonObjectSchema(generateJsonSchemaForTypedescAsJson(td));
 }
+
+isolated function generateJsonObjectSchema(map<json> schema) returns SchemaResponse {
+    string[] supportedMetaDataFields = ["$schema", "$id", "$anchor", "$comment", "title", "description"];
+
+    if schema["type"] == "object" {
+        return {schema};
+    }
+
+    map<json> updatedSchema = map from var [key, value] in schema.entries()
+        where supportedMetaDataFields.indexOf(key) is int
+        select [key, value];
+
+    updatedSchema["type"] = "object";
+    map<json> content = map from var [key, value] in schema.entries()
+        where supportedMetaDataFields.indexOf(key) !is int
+        select [key, value];
+
+    updatedSchema["properties"] = {[RESULT]: content};
+
+    return {schema: updatedSchema, isOriginallyJsonObject: false};
+}
+
+isolated function getGetLlmResultTool(map<json> parameters) returns ai:ChatCompletionFunctions[]|error {
+    return [{
+        name: GET_RESULTS_TOOL,
+        parameters: check parameters.toJson().cloneWithType(),
+        description: string 
+            `Tool to call with the response from a large language model (LLM) for a user prompt.`
+    }];
+}
+
+isolated function getToolChoiceToGenerateLlmResult() returns ChatCompletionNamedToolChoice => {
+        'function: {
+            name: GET_RESULTS_TOOL
+        }
+    };
+
+isolated function generateOpenAIChatCompletionTools(ai:ChatCompletionFunctions[] tools) returns ChatCompletionTool[] =>
+         from ai:ChatCompletionFunctions tool in tools select {
```

Comment:
Since this is only used internally, we can check for an array of length 1 instead of using a query expression.